### DISCUSSION TITLE
When a new image is built, kargo will generate two freights, which will cause some problems

### DISCUSSION THREAD

**Discussion by wokwong** [akuity/kargo#2948] [Q&A] [answered]
When a new image is built and uploaded to the image registry, kargo generates two items. Details as follows：
My warehouse subscribes to：
- a chart registry
- an image registry
- a git repo（for applications' values.yaml）

Assume that the current chart version is unchanged, the image version is `v0.3`, and the values git reop commitID is `8e5ee3`:
1. now build a new image version for `v0.4`, then kargo will generate the freight  `dangling-quetzal` (8e5ee3+v0.4), 
2. then, promoting the freight will update the values git reop with a commitID of `35a09c`, kargo will generate another new freight `cranky-gopher` (35a09c+v0.4).

![image](https://github.com/user-attachments/assets/1fc02103-bce8-4815-b3cc-8b3b5222fda0)

Here's the problem: When promoting `dangling-quetzal`, argocd's `LAST SYNC` status shows that it has been refreshing application frequently,and the annotations of the argo application change repeatedly. When `cranky-gopher` was promoted, all was well. Why is that?
```bash
$ kubectl get applications.argoproj.io -n argocd wj-proj-dev -o=jsonpath='{.metadata.annotations}' -w
{"argocd.argoproj.io/refresh":"hard","kargo.akuity.io/authorized-stage":"wj-proj-kargo:dev"}
{"kargo.akuity.io/authorized-stage":"wj-proj-kargo:dev"}
{"kargo.akuity.io/authorized-stage":"wj-proj-kargo:dev"}
{"kargo.akuity.io/authorized-stage":"wj-proj-kargo:dev"}
{"argocd.argoproj.io/refresh":"hard","kargo.akuity.io/authorized-stage":"wj-proj-kargo:dev"}
{"argocd.argoproj.io/refresh":"hard","kargo.akuity.io/authorized-stage":"wj-proj-kargo:dev"}
{"argocd.argoproj.io/refresh":"hard","kargo.akuity.io/authorized-stage":"wj-proj-kargo:dev"}
```

**Comment by krancour** [comment, marked as answer]
> then, promoting the freight will update the values git reop with a commitID of 35a09c, kargo will generate another new cargo cranky-gopher (35a09c+v0.4).

If your promotion process is writing back to the same branch to which your Warehouse subscribes, then your promotion process is indirectly creating the second piece of Freight.

Kargo used to not let you do this (for the exact reason you are now encounting), but so many different features have emerged since then that can help you to mitigate this, so that protection eventually vanished.

1. You can use path filters on your Warehouse's git repo subscription to try and ignore new commits where the only changes are to certain files.

1. You can disable the Warehouse from automatically creating Freight, in which case it will still discover new artifacts and it is up to you to choose artifacts from a menu to create new Freight.

1. Recommended: Don't have your promotion process write back to the same branch you're subscribed to. Write to a stage-specific branch instead. This dramatically simplifies _everything_.

**Reply by wokwong** [reply]
OK, thanks for your reply. 

> When promoting dangling-quetzal, argocd's LAST SYNC status shows that it has been refreshing application frequently,and the annotations of the argo application change repeatedly. 

I still don't understand why，what is the reason？

**Reply by krancour** [reply]
I would need to look at all configs in much greater detail, but I can take a _guess_ that the App is syncing to the HEAD of your main branch. When Kargo tells the App to sync at the tail end of `dangling-quetzal`, it's watching/waiting for it to finish syncing to 8e5ee3, but instead keeps seeing that it's synced to 35a09c... so it's telling it to sync again.

This is a guess based on limited information.

I do _strongly_ recommend, however, not writing back to a branch you subscribe to.

I like to think of `main` as "source" and my promotion process as being analogous to a compilation process. Store the result somewhere else. Don't overwrite the source.

**Reply by krancour** [reply]
@wokwong, I just wanted to circle back and let you know that #2980, which will be included in the upcoming v1.1.0 release will likely permit you to do what you were initially attempting to do.
